### ISSUE TITLE
Generic support for the Jupyter display protocol in the JLab terminal

### ISSUE THREAD

**Issue by fperez** [jupyterlab/jupyterlab#9662]
### Problem

The inspiration for this came from #5056, but during today's governance call we brainstormed a slightly more generic solution... 

The problem is how to generically allow operations at the JLab command line (terminal widget) to perform general actions in JupyterLab, of which a hypothetical `jlopen` command would be one (albeit very important) case.

### Proposed Solution

We were thinking about using the Jupyter display protocol as the generic mechanism, akin to how in the notebook we have stdout, stderr and rich output with our mime bundle payloads. @blink1073 indicated we could capture with some custom control codes output at the terminal, and the payload after such a special code could be defined as a rich Jupyter output.  

This model would enable a generic `jlopen`, but others could implement new utilities to display/manipulate data in JupyterLab from the terminal.

cc @blink1073 @ellisonbg @lresende @afshin

**Comment by jasongrout** [comment]
The display protocol also supports updating visually-displayed material, fallbacks, etc., and is customized around the needs of *displaying* something, which IIRC is not really our usecase here. Perhaps something simpler is needed here, just a generic control protocol.

**Comment by ellisonbg** [comment]
Yep, this leverages a subset of the display system, namely the MIME bundle part of it and the rendermime interfaces in JupyterLab (doesn't use the actual display protocol, the transient field, etc.). The implementation uses the existing display objects from `IPython` so it does include fallbacks (multiple MIME types).

**Comment by jasongrout** [comment]
> The implementation uses the existing display objects from `IPython` so it does include fallbacks (multiple MIME types).

There is a presumption here that the display app bundled with jlab is the one that is run in the terminal. If jlab is started from an environment, but the terminal is not, that assumption can pretty easily fail (the display app won't exist, or possibly will be from a different environment and different version, etc.).

I'm not sure a clean way around this, unless terminals are always started in the same environment that the server was started in. Is that the case?

It may be that there is nothing we can really do automatically about this, in which case we should document this.

**Comment by pdeffebach** [comment]
Bumping this, as I think it is the only open issue discussing this issue. It would be great to have a `jlopen` command in jupyterlab which allows me to edit a text file that I open from the terminal.

**Comment by TidbitSoftware** [comment]
A related ask might be a command called, perhaps, `jlcd` which would allow the user to change the directory displayed in the File Browser. Use case might be that they have found an interesting notebook that they want to run in directory `foo`. They could then run `jpcd foo` to change the displayed directory in the File Browser, then launch the notebook via the GUI.

**Comment by fperez** [comment]
@TidbitSoftware - good point, though implementation-wise, I'd suggest subsuming that into the single "open" command, so that `jopen <path>` opens that path in the file browser if the path is a directory. This is how MacOS does it and I think it's a good model, `open foo` will trigger the Finder if `foo` is a directory.  That way users only have to worry about one command.

A separate UX question is how to better offer file management outside of the single, fixed, global left pane tool. But that deserves its own thread :)

**Comment by josephko91** [comment]
Is there an update on this by any chance? It seems like an incredibly useful and practical command to have within the jlab terminal.

**Comment by fperez** [comment]
Tagging @dlqqq re. discussion at jupyter-ai dev summit for reference on `jl-cli` ideas.

**Comment by jtpio** [comment]
Linking to https://github.com/jupyterlite/terminal/pull/78 as related to this idea, where @ianthomas23 started looking into being able to execute JupyterLab commands from the terminal.

There is also some prior work by @bollwyvl.

Mentioning that here since the display protocol (and other common tasks like opening files) may just be a subset of actions that could be also achieved with a more generic command based approach.

**Comment by fperez** [comment]
Agreed @jtpio! And indeed, the discussion we had with @dlqqq at the jupyter-ai SF meeting last week was really about exposing a complete, and comprehensible, set of CLI entry points to control JupyterLab.

@dlqqq referred to [cmux](https://cmux.com) as inspiration for a well-rounded CLI to control the entire application. I haven't used it myself but I'm curious.

**Comment by jtpio** [comment]
> [@dlqqq](https://github.com/dlqqq) referred to [cmux](https://cmux.com) as inspiration for a well-rounded CLI to control the entire application. I haven't used it myself but I'm curious.

Interesting thanks for sharing. `cmux` looks similar to the JupyterLab based metapackage I've been experimenting with lately (https://github.com/jtpio/xtralab), which uses `jupyter-server-mcp` and `jupyterlab-commands-toolkit` and makes it possible for existing coding agents to "run arbitrary JupyterLab commands in JupyterLab".

> There is also some prior work by @bollwyvl.

This one: https://github.com/deathbeds/jyg. It looks quite aligned with the idea of such `jl-cli` tool.

**Comment by fperez** [comment]
Thx @jtpio for those - glad to see this discussion reactivating. I hope this converges :)
